src/ai/backend/storage/filebrowser/filebrowser.py

```diff
@@ -0,0 +1,190 @@
+from __future__ import annotations
+
+import asyncio
+import logging
+from datetime import datetime
+
+import aiodocker
+
+from ai.backend.common.logging import BraceStyleAdapter
+from ai.backend.common.validators import BinarySize
+from ai.backend.storage.context import Context
+from ai.backend.storage.utils import (
+    get_available_port,
+    is_port_in_use,
+    mangle_path,
+)
+
+from .config_browser_app import prepare_filebrowser_app_config
+from .database import SQLite_DB
+
+log = BraceStyleAdapter(logging.getLogger(__name__))
+
+__all__ = (
+    "create_or_update",
+    "destroy_container",
+    "get_container_by_id",
+    "get_filebrowsers",
+    "get_network_stats",
+)
+
+
+async def create_or_update(ctx: Context, vfolders: list[dict]) -> tuple[str, int, str]:
+    image = ctx.local_config["filebrowser"]["image"]
+    service_ip = ctx.local_config["filebrowser"]["service-ip"]
+    service_port = ctx.local_config["filebrowser"]["service_port"]
+    max_containers = ctx.local_config["filebrowser"]["max-containers"]
+    cgroup = ctx.local_config["filebrowser"]["cgroup"]
+    settings_path = ctx.local_config["filebrowser"]["settings_path"]
+    mount_path = ctx.local_config["filebrowser"]["mount_path"]
+    cpu_count = ctx.local_config["filebrowser"]["max-cpu"]
+    memory = ctx.local_config["filebrowser"]["max-mem"]
+    memory = int(BinarySize().check_and_return(memory))
+    db_path = ctx.local_config["filebrowser"]["db-path"]
+
+    if is_port_in_use(service_port):
+        service_port = get_available_port()
```

Comment:
Let's rely on Docker's feature on port-range to automatically assign available port number in a given range of ports.

Also, we need to make the port range **configurable** because we should be able to sync with the customer site firewall configs.

Example:
<img width="1389" alt="image" src="https://user-images.githubusercontent.com/555156/161217524-70747037-13ed-40f5-92b4-3de706ccb093.png">

Reply:
Implemented the setting the range of available port numbers. Which is set in storage-proxy.toml

However, using docker api it is not possible to automatically check by docker the port availablility. "4000-5000" syntax doesn't work when value assigned in json.

Exposed port is the port which will be open in docker container and the filebrowser will be listening on this port.
In HostConfig, we bind the container exposed port to HostPort which is determined before the creating container through docker api.
```
 "ExposedPorts": {
                "8080/tcp": {},
            },
            "Image": image,
            "HostConfig": {
                "PortBindings": {
                    "8080/tcp": [
                        {
                            "HostIp": f"{service_ip}",
                            "HostPort": f"{service_port}/tcp",
                        },
                    ],
                },
```